SignalServiceKit/src/Messages/OWSMessageManager.m

```diff
@@ -1053,6 +1070,70 @@ - (TSIncomingMessage *_Nullable)handleReceivedEnvelope:(OWSSignalServiceProtosEn
     }
 }
 
+- (TSQuotedMessage *_Nullable)quotedMessageForDataMessage:(OWSSignalServiceProtosDataMessage *)dataMessage
+{
+    OWSAssert(dataMessage);
+
+    if (!dataMessage.hasQuote) {
+        return nil;
+    }
+
+    OWSSignalServiceProtosDataMessageQuote *quoteProto = [dataMessage quote];
+
+    if (![quoteProto hasId] || [quoteProto id] == 0) {
+        OWSFail(@"%@ quoted message missing id", self.logTag);
+        return nil;
+    }
+    uint64_t timestamp = [quoteProto id];
+
+    if (![quoteProto hasAuthor] || [quoteProto author].length == 0) {
+        OWSFail(@"%@ quoted message missing author", self.logTag);
+        return nil;
+    }
+    // TODO: We could verify that this is a valid e164 value.
+    NSString *authorId = [quoteProto author];
+
+    NSString *_Nullable body = nil;
+    BOOL hasText = NO;
+    BOOL hasAttachment = NO;
+    if ([quoteProto hasText] && [quoteProto text].length > 0) {
+        body = [quoteProto text];
+        hasText = YES;
+    }
+
+    NSString *_Nullable sourceFilename = nil;
+    NSData *_Nullable thumbnailData = nil;
+    NSString *_Nullable contentType = nil;
+
+    if ([quoteProto hasAttachment]) {
+        OWSSignalServiceProtosAttachmentPointer *attachmentProto = [quoteProto attachment];
+        if ([attachmentProto hasContentType] && attachmentProto.contentType.length > 0) {
+            contentType = attachmentProto.contentType;
+
+            if ([attachmentProto hasFileName] && attachmentProto.fileName.length > 0) {
+                sourceFilename = attachmentProto.fileName;
+            }
+            if ([attachmentProto hasThumbnail] && attachmentProto.thumbnail.length > 0) {
```

Comment:
This is safe to keep in, but for the record, I don't think any clients are going to be sending raw thumbnail data for now.

We should instead:

Use (duplicate?) the quoted attachment in the likely even that the message being quoted has it's attachment around.

Else download the attachment referenced by this pointer.

Reply:
I'll make a card.